alternator/conditions.cc

```diff
@@ -131,6 +132,37 @@ static bool check_BEGINS_WITH(const rjson::value* v1, const rjson::value& v2) {
     return val1.substr(0, val2.size()) == val2;
 }
 
+// Check if a JSON-encoded value equals any element of an array.
+static bool check_IN(const rjson::value* val, const rjson::value& array) {
+    if (array.Size() < 1) {
+        throw api_error("ValidationException", "IN operator requires nonempty AttributeValueList");
+    }
+    if (!array[0].IsObject() || array[0].MemberCount() != 1) {
+        throw api_error("ValidationException",
+                        "IN operator requires AttributeValueList elements to be of type String, Number, or Binary ");
```

Comment:
This error message is misleading and confused me. The problem here is that the item is malformed - doesn't even conform to the way DynamoDB represents typed objects - it's not even about string, number or binary specifically.